File: mancala.py
class board:

    def __init__(self):
        self.board = [[4,4,4,4,4,4,0],[4,4,4,4,4,4,0]]

    def move(self, side, place):
        handfull = self.board[side][place]
        self.board[side][place] = 0
        playerside = side
        place += 1
        while handfull:
            if place < 6 or (playerside == side and place == 6): 
                self.board[side][place] += 1
                handfull -= 1
                place += 1
            else:
                side = (side + 1)%2
                place = 0

        if place < 7 and self.board[side][place - 1] == 1 and playerside == side:
            self.board[playerside][6] += self.board[side][place - 1] + self.board[(side + 1)%2][6 - place]
            self.board[side][place - 1] = 0
            self.board[(side + 1)%2][6-place] = 0

        if place == 7 and side == playerside:
            return playerside
        else:
            return (playerside + 1)%2

    def check_winner(self):
        empty1 = 0
        empty2 = 0
        gameover = False
        for i in range(len(self.board[0])-1):
            empty1+=self.board[0][i]
            empty2+=self.board[1][i]

        #check for game_over
        if not empty1:
            self.board[1][6] += empty2
            gameover = True
        if not empty2:
            self.board[0][6] += empty1
            gameover = True

        if gameover:
            if self.board[0][6] < self.board[1][6]:
                return 2
            else:
                return 1

        return 0
            

    def printboard(self):
        side1 = '  '
        side2 = '  '
        ends = str(self.board[1][6])
        for i in range(len(self.board[0])-1):
            side1 += str(self.board[0][i]) + " "
            side2 += str(self.board[1][5-i]) + " "
            ends += "  "
        ends += " " + str(self.board[0][6])
        print side1
        print ends
        print side2

def main():
    game = board()
    side = 0
    winner = 0
    while not winner:
        game.printboard()
        place = int(raw_input("player " + str(side + 1) + " to move... which space to move? ")) - 1
        side = game.move(side, place)
        winner = game.check_winner()
    print "player " + str(winner) + " wins!"
        

if __name__ == '__main__':
    main()

File: mancala.c
#include<stdio.h>
#include<string.h>

#define LENGTH 3

struct Vertex;

typedef struct Vertex Vertex;

struct Vertex {int sides[2][LENGTH]; int ends[2]; Vertex *children; Vertex *parent;};

int has_more(struct Vertex *node, int side) {
    int i;
    int more = 0;
    for (i=0;i<LENGTH;i++)
        more += (*node).sides[side][i];
    return more;
} 

/* gets a winner when a node is end of game */
int get_winner(struct Vertex *node) {
    /*load all pieces into ends*/
    int i;
    int j;
    for (j=0;j<2;j++)
        for (i=0;i<LENGTH;i++) {
            (*node).ends[j] += (*node).sides[j][i];
            (*node).sides[j][i] = 0;
        }

        

    if ((*node).ends[0] < (*node).ends[1])
        return 1;
    else if ((*node).ends[1] < (*node).ends[0])
        return 0;
    else
        return -1; 
}

/* determines if a player is a definite winner at intermediate stage.  The current player just needs ONE child to be a definite win.  The opposing player needs ALL children to be definite wins */
int winning_player(int children_wins[LENGTH], int side) {
    int opposingside = (side + 1)%2;
    int i;
    /*if player can force a win, he forces a win*/
    for (i = 0; i < LENGTH; i++)
        if (children_wins[i] == side)
            return side;
    /*if player cannot win, but can force a tie, he forces a tie*/
    for (i = 0; i < LENGTH; i++)
        if (children_wins[i] == -1)
            return -1;
    /*if player cannot force a win or force a tie, then the opposing player must win*/
    return opposingside;
}
    
/*returns whose move it is now*/
int move(Vertex *node, int side, int place) {

    int playerside = side;
    int opposingside = (side + 1)%2;
    int handfull = (*node).sides[side][place];
    (*node).sides[side][place] = 0;
    place += 1;

    while (handfull) {
        if ((place+1)%(LENGTH+1) ) {
            (*node).sides[side][place] += 1;
            place += 1;
            handfull -= 1;
        } else if (playerside == side) {
            (*node).ends[side] += 1;
            place = 0;
            side = opposingside;
            handfull -= 1;
        } else {
            place = 0;
            side = playerside;
        }
    }

    /*if ending in an empty spot, take yours and all across from it*/
    if (playerside == side && place != 0)
        if ((*node).sides[side][place-1] == 1) {
            (*node).ends[side] += (*node).sides[side][place-1] + (*node).sides[opposingside][LENGTH - place];
            (*node).sides[side][place-1] = 0;
            (*node).sides[opposingside][LENGTH - place] = 0;
        }

    /*if ending in own end, go again*/
    if (side == opposingside && place == 0)
        return playerside; 
    else
        return opposingside;
}

int is_full(Vertex *node, int side, int place) {
    if ((*node).sides[side][place])
        return 1;
    else
        return 0;
}

Vertex init_child(Vertex *node, Vertex *children) {
    int i;
    int j;
    Vertex child = (*node);
    for (j = 0; j < 2; j++) {
        for (i = 0; i < LENGTH; i++)
            child.sides[j][i] = (*node).sides[j][i];
        child.ends[j] = (*node).ends[j];
    }
    child.children = children;
    child.parent = node;
    return child;
}

void print_board(Vertex *node) {
    int i;
    printf("\n\n  ");
    for (i = 0; i<LENGTH; i++)
        printf("%d ", (*node).sides[0][LENGTH -1 -i]);
    printf("\n%d ", (*node).ends[0]);
    for (i=0; i<LENGTH;i++)
        printf("  ");
    printf("%d\n  ", (*node).ends[1]);
    for (i = 0; i<LENGTH; i++)
        printf("%d ", (*node).sides[1][i]);
    printf("\n");
    return;
}


/* This returns whether this child is a definite win and for which side. -1 = no definite win, 0 = side 0 win, 1 = side 1 win */
int depth_first(Vertex *node, int side) {
    int i;
    int opposingside = (side + 1)%2;

    int child_wins[LENGTH];
    for (i=0;i<LENGTH;i++)
        child_wins[i] = -2;

    int winner = -2;
    int nextplayer;
    /*end of game*/
    if (!has_more(node, opposingside) || !has_more(node, side)) {
        winner = get_winner(node);
        printf("\n============================");
        printf("\n%d's turn, %d wins! \n", side, winner);
        print_board(node);
        printf("============================\n");
        return winner;
    }

    /*recurse into all possible moves*/
    for (i = 0; i < LENGTH; i++) {
        if (is_full(node, side, i)) {
            /*initializes a child to be just like parent, then moves the child*/  
            Vertex children[LENGTH];
            *((*node).children + i) = init_child(node, children);
            nextplayer = move((*node).children + i, side, i);
            child_wins[i] = depth_first((*node).children + i, nextplayer);
        }
    }

    /*figure out who can win with a perfect strategy at this point*/
    winner = winning_player(child_wins, side);
    printf("\n---------------------------------------");
    printf("\n%d's turn, player %d is winning!\n", side, winner);
    for (i=0;i<LENGTH;i++)
        printf("%d ", child_wins[i]);
    print_board(node);
    printf("---------------------------------------\n");
    return winner; 
}
                
int main(int argc, char **argv) {

    printf("what?");
    int winner;
    Vertex children[LENGTH];
    Vertex head = {.sides = {{4,4,4}, {4,4,4}}, .ends = {0}, .children = children, .parent = NULL};
    winner = depth_first(&head, 1);
    printf("\nside %d has a winning strategy!", winner);
    return 0;
}
    
    
    
    
    

